#pragma once

#include "ops/linear/common/output.cuh"

namespace ninfer::ops::detail {

// Thread-local operations over a fully reduced accumulator. The contraction
// owns predicates and synchronization; epilogues only receive valid coordinates.
struct LinearIdentityEpilogue {
    __device__ __forceinline__ float apply(int, int, float value) const { return value; }

    __device__ __forceinline__ float apply_scaled(int, int, float value, float scale) const {
        return value * scale;
    }
};

struct LinearResidualAddEpilogue {
    LinearBf16InputView residual;

    __device__ __forceinline__ float apply(int row, int token, float value) const {
        return value + residual.load(row, token);
    }

    // Scaling belongs to the fused residual update, and it is rounded as a product and then a sum in
    // every tile. Whether the caller's live-column predicate is compile-time must not decide the
    // rounding: it used to, because the multiply and the add were separate statements, so only
    // launches whose width filled every token tile let the compiler contract them into one FMA. Those
    // widths then rounded differently from all the others, and the same token prefilled to different
    // bits depending on where it fell in the chunk.
    __device__ __forceinline__ float apply_scaled(int row, int token, float value,
                                                  float scale) const {
        return __fadd_rn(__fmul_rn(value, scale), residual.load(row, token));
    }
};

// Row consumers run in the single thread owning the fully reduced row. They
// may publish fused Op state, but do not introduce warp/CTA synchronization.
template <class Output, class Epilogue, int Tokens>
__device__ __forceinline__ void
linear_finish_row(const Output& output, const Epilogue& epilogue, int row, int token_begin,
                  const float (&values)[Tokens], int active_tokens) {
    if constexpr (requires {
                      epilogue.apply_row(output, row, token_begin, values, active_tokens);
                  }) {
        epilogue.apply_row(output, row, token_begin, values, active_tokens);
    } else {
#pragma unroll
        for (int token = 0; token < Tokens; ++token) {
            if (token < active_tokens) {
                const int global_token = token_begin + token;
                output.store(row, global_token, epilogue.apply(row, global_token, values[token]));
            }
        }
    }
}

template <class Schedule, class Epilogue>
inline constexpr int linear_epilogue_bytes = [] {
    if constexpr (requires { Epilogue::template kSharedBytes<Schedule>; })
        return static_cast<int>(Epilogue::template kSharedBytes<Schedule>);
    else
        return 0;
}();

template <class Schedule, class Epilogue>
inline constexpr int linear_mma_shared_bytes =
    Schedule::kSharedBytes > linear_epilogue_bytes<Schedule, Epilogue>
        ? Schedule::kSharedBytes
        : linear_epilogue_bytes<Schedule, Epilogue>;

} // namespace ninfer::ops::detail
